### ISSUE TITLE
dotnet publish large use ram

### ISSUE THREAD

**Issue by Igor-ZA** [dotnet/sdk#31259]
### Description

When creating an APK, the dotnet publish command uses a lot of memory.
![image](https://user-images.githubusercontent.com/56849652/225612981-da017c31-48c7-47a8-876b-9f5925845587.png)

### Steps to Reproduce

1. Create a File > New .NET MAUI Blazor App
2. Add package Microsoft.EntityFrameworkCore.Sqlite 7.0.4
3. App folder: dotnet publish .\MauiApp1.csproj -f:net7.0-android -c:Release /p:PublishTrimmed=true

### Link to public reproduction project repository

https://github.com/Igor-ZA/MauiApp1.git

### Version with bug

7.0 (current)

### Last version that worked well

7.0 (current)

### Affected platforms

Android

### Affected platform versions

Android

**Comment by nagilson** [comment]
Do we have data on if this is a regression? Not sure there's much to do here. Personally, I would rather my publish use a lot of my ram if it's available so it can get the job done faster.

**Comment by marcpopMSFT** [comment]
@jonathanpeppers @agocke @lewing not sure if this is related to blazor, maui, or trimming. I don't think it's likely to be in the SDK as publishing logic is fairly simple. Is it large files all getting loaded into memory from the build and then copied potentially?

**Comment by jonathanpeppers** [comment]
What are the `dotnet.exe` processes? Are they instances of MSBuild? I would expect the AOT compiler to show `mono-aot-compiler.exe`, so this is odd.

@Igor-ZA are all the repro steps required?

> * Create a File > New .NET MAUI Blazor App
> * Add package Microsoft.EntityFrameworkCore.Sqlite 7.0.4
> * App folder: dotnet publish .\MauiApp1.csproj -f:net7.0-android -c:Release /p:PublishTrimmed=true

It only happens with MAUI+Blazor *and* entity framework is included?

**Comment by steveisok** [comment]
The binlog would be helpful to see where it's hanging out the most.

**Comment by Igor-ZA** [comment]
@jonathanpeppers Hi!

I see that the publishing process uses a lot of memory by default, but the more packages I add to it, the more memory it will use.
I only looked at MAUI Blazor.

**Comment by jonathanpeppers** [comment]
We run an "inner" build per RuntimeIdentifier (default is 4):

https://github.com/xamarin/xamarin-android/blob/30e94872bf9dc328941d0908080ee25b10b19c52/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AssemblyResolution.targets#L103-L108

But I would be curious if the `dotnet.exe` is actually MSBuild or the trimmer.

**Comment by Igor-ZA** [comment]
I don't know. :(
How can I find out?

**Comment by jonathanpeppers** [comment]
Easiest is probably the command-line column you can add on the Details tab:

![image](https://user-images.githubusercontent.com/840039/227040529-87284f2f-dcb2-4c64-b686-a39f8c7c57c2.png)

On this example, it is Roslyn, `dotnet exec VSCSCompiler.dll`. Yours might be `ILLink.dll` or `MSBuild.dll`.

**Comment by Igor-ZA** [comment]
![image](https://user-images.githubusercontent.com/56849652/227043318-3e2fa729-3588-40de-9028-e47f73af88e5.png)

**Comment by jonathanpeppers** [comment]
/cc @sbomer

This app is probably loading 100+ assemblies, if it is MAUI + Blazor + Entity Framework.

I'm guessing it just uses this much RAM loading them all?

https://github.com/dotnet/runtime/blob/1d9c93f729388c443d47843e6d683c2625b2c901/src/tools/illink/src/linker/Linker/AssemblyResolver.cs#L141-L147

**Comment by agocke** [comment]
Yeah, this is not an unusual amount of memory to trim 100+ assemblies. I think the main question is why multiple apps are being trimmed in parallel.

**Comment by jonathanpeppers** [comment]
It’s because an Android app has 4 architectures by default. We may be able to drop 32 bit in the .NET 8 timeframe.

**Comment by agocke** [comment]
I see -- in that case it might be appropriate to build each architecture one-by-one. That would limit memory usage, at the expense of increased time.

**Comment by jonathanpeppers** [comment]
There is a undocumented setting `<_AndroidBuildRuntimeIdentifiersInParallel>false</_AndroidBuildRuntimeIdentifiersInParallel>` that would opt out of this.
